flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/reuse/InputPriorityConflictResolver.java

```diff
@@ -0,0 +1,353 @@
+		Preconditions.checkState(
+			node.getInputNodes().size() == node.getInputEdges().size(),
+			"Number of inputs nodes does not equal to number of input edges for node " +
+				node.getClass().getName() + ". This is a bug.");
+		for (int i = 0; i < node.getInputNodes().size(); i++) {
+			int priority = node.getInputEdges().get(i).getPriority();
+			inputPriorityGroupMap.computeIfAbsent(priority, k -> new ArrayList<>()).add(i);
+		}
+
+		// add edges between neighboring priority groups
+		List<List<Integer>> inputPriorityGroups = new ArrayList<>(inputPriorityGroupMap.values());
+		for (int i = 0; i + 1 < inputPriorityGroups.size(); i++) {
+			List<Integer> higherGroup = inputPriorityGroups.get(i);
+			List<Integer> lowerGroup = inputPriorityGroups.get(i + 1);
+
+			for (int higher : higherGroup) {
+				for (int lower : lowerGroup) {
+					addTopologyEdges(node, higher, lower);
+				}
+			}
+		}
+	}
+
+	private void addTopologyEdges(ExecNode<?, ?> node, int higherInput, int lowerInput) {
+		ExecNode<?, ?> higherNode = node.getInputNodes().get(higherInput);
+		ExecNode<?, ?> lowerNode = node.getInputNodes().get(lowerInput);
+		List<ExecNode<?, ?>> lowerAncestors = calculateAncestors(lowerNode);
+
+		List<Tuple2<ExecNode<?, ?>, ExecNode<?, ?>>> linkedEdges = new ArrayList<>();
+		for (ExecNode<?, ?> ancestor : lowerAncestors) {
+			if (graph.link(higherNode, ancestor)) {
+				linkedEdges.add(Tuple2.of(higherNode, ancestor));
+			} else {
+				// a conflict occurs, resolve it by adding a batch exchange
+				// and revert all linked edges
+				if (lowerNode instanceof BatchExecExchange) {
+					BatchExecExchange exchange = (BatchExecExchange) lowerNode;
+					exchange.setRequiredShuffleMode(ShuffleMode.BATCH);
+				} else {
+					node.replaceInputNode(lowerInput, (ExecNode) createExchange(node, lowerInput));
+				}
+
+				for (Tuple2<ExecNode<?, ?>, ExecNode<?, ?>> linkedEdge : linkedEdges) {
+					graph.unlink(linkedEdge.f0, linkedEdge.f1);
+				}
+				return;
+			}
+		}
+	}
+
+	/**
+	 * Find the ancestors by going through PIPELINED edges.
+	 */
+	@VisibleForTesting
+	List<ExecNode<?, ?>> calculateAncestors(ExecNode<?, ?> node) {
+		List<ExecNode<?, ?>> ret = new ArrayList<>();
+		AbstractExecNodeExactlyOnceVisitor ancestorVisitor = new AbstractExecNodeExactlyOnceVisitor() {
+			@Override
+			protected void visitNode(ExecNode<?, ?> node) {
+				List<ExecEdge> inputEdges = node.getInputEdges();
+				boolean hasAncestor = false;
+
+				for (int i = 0; i < inputEdges.size(); i++) {
+					// we only go through PIPELINED edges
+					if (inputEdges.get(i).getDamBehavior().stricterOrEqual(ExecEdge.DamBehavior.END_INPUT)) {
+						continue;
+					}
+					hasAncestor = true;
+					node.getInputNodes().get(i).accept(this);
+				}
+
+				if (!hasAncestor) {
+					ret.add(node);
+				}
+			}
+		};
+		node.accept(ancestorVisitor);
+		return ret;
+	}
+
+	private BatchExecExchange createExchange(ExecNode<?, ?> node, int idx) {
```

Comment:
could you provide more test cases to cover different branches ?

Reply:
Existing tests in `DeadlockBreakupTest` already covers the possible branches.